src/shared/tpm2-util.c

```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary);
+                if (rc)
+                        return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                                "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                                sym_Tss2_RC_Decode(rc));
+                break;
+        }
+
+        /* Did Not find SRK, indicate this by returning 0 and ESYS_TR_NONE */
+        if (primary == ESYS_TR_NONE) {
+                *ret_primary = ESYS_TR_NONE;
+                return 0;
+        }
+
+        /* Found the SRK, make sure the alg is what we were expecting */
+        _cleanup_(Esys_Freep) TPM2B_PUBLIC *out_public = NULL;
+        _cleanup_(Esys_Freep) TPM2B_NAME *out_name = NULL, *out_qname = NULL;
+        rc = sym_Esys_ReadPublic(
+            c,
+            primary,
+            ESYS_TR_NONE,
+            ESYS_TR_NONE,
+            ESYS_TR_NONE,
+            &out_public,
+            &out_name,
+            &out_qname);
+        if (rc)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Cannot specify the alg and have a mismatch */
+        if (alg != 0 && alg != out_public->publicArea.type)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "SRK expected alg 0x%x does not match found alg 0x%x",
+                                        alg, out_public->publicArea.type);
+
+        *ret_primary = primary;
+
+        if (ret_alg)
+                 *ret_alg = out_public->publicArea.type;
+
+        return 0;
+}
+
+
+static int tpm2_make_primary(
+                ESYS_CONTEXT *c,
+                ESYS_TR *ret_primary,
+                TPMI_ALG_PUBLIC alg,
+                bool use_srk_model,
+                TPMI_ALG_PUBLIC *ret_alg) {
+
+        static const TPM2B_SENSITIVE_CREATE primary_sensitive = {};
+        const TPM2B_PUBLIC *primary_template = NULL;
+
         static const TPML_PCR_SELECTION creation_pcr = {};
         ESYS_TR primary = ESYS_TR_NONE;
         TSS2_RC rc;
         usec_t ts;
 
-        log_debug("Creating primary key on TPM.");
+        /* Find existing SRK and use it if present */
+        if (use_srk_model) {
+                rc = tpm2_get_srk(c, ret_primary, alg, ret_alg);
+                if (rc)
+                        return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                               "Failed to establish if SRK is present");
+
+                if (*ret_primary != ESYS_TR_NONE) {
+                        log_debug("Discovered SRK using that");
+                        return 0;
+                }
+                log_debug("Did not find SRK, generating...");
```

Comment:
hmm, how does tpm2 access locking actually work? i.e. if multiple instances of this code run in parallel, will they just work? or will possibly both detect that an SRK is missing, and both create it in parallel, and once wins and the othe rfails?

Reply:
Is this holding /dev/tpmrm0? if so it's per-command transaction, if it's holding /dev/tpm0, then it's exclusive.

Reply:
But either way if this code can run in parallel it will need a lock here, whats the mechanism for that in systemd?

Reply:
it's /dev/tpmrm0.

So I did get this right: it's one step to check if the SRK is existant, and another one to then generate it, and we need to manually lock around it?

note that cryptsetup instances for different volumes run entirely in parallel. so if there's stuff they should not do in parallel, we need a system-wide mutex for that. Typically we use "unposix" or BSD locks for that (since they are per-fd, auto-cleaning and robust) on some suitable file. This could be the /dev/tpmrm0 fd itself actually, but not sure if the tss2 libs allow us access to the fd it internally uses. Also, it might be too coarse to just say "lock the whole device for everyone", and we'd much rather have a more finer grained lock for just "lock the init srk operation only".

assuming we want the latter, you could use the existing make_lock_file() helper, with a path such as "/run/systemd/tpm2-srk-init" or so.

Reply:
> it's /dev/tpmrm0.
> 
> So I did get this right: it's one step to check if the SRK is existant, and another one to then generate it, and we need to manually lock around it?
> 

Yes, 100% correct.

> note that cryptsetup instances for different volumes run entirely in parallel. so if there's stuff they should not do in parallel, we need a system-wide mutex for that. Typically we use "unposix" or BSD locks for that (since they are per-fd, auto-cleaning and robust) on some suitable file. This could be the /dev/tpmrm0 fd itself actually, but not sure if the tss2 libs allow us access to the fd it internally uses. 

GetPollHandles

> Also, it might be too coarse to just say "lock the whole device for everyone", and we'd much rather have a more finer grained lock for just "lock the init srk operation only".

Yes too coarse IMO.

> assuming we want the latter, you could use the existing make_lock_file() helper, with a path such as "/run/systemd/tpm2-srk-init" or so.

Perfect, thanks.